File: sdks/typescript/src/team_strategy_values.ts
/** Team collaboration strategies projected without owning Team execution. */
export type TeamStrategy =
  | { readonly kind: "manager_subagent" }
  | { readonly kind: "pipeline"; readonly members: readonly string[] }
  | { readonly kind: "debate"; readonly judge: string; readonly debaters: readonly string[] }
  | { readonly kind: "swarm"; readonly reducer: string };

export function teamStrategyManager(): TeamStrategy {
  return Object.freeze({ kind: "manager_subagent" });
}

export function teamStrategyPipeline(members: readonly string[]): TeamStrategy {
  return Object.freeze({ kind: "pipeline", members: Object.freeze([...members]) });
}

export function teamStrategyDebate(judge: string, debaters: readonly string[]): TeamStrategy {
  return Object.freeze({ kind: "debate", judge, debaters: Object.freeze([...debaters]) });
}

export function teamStrategySwarm(reducer: string): TeamStrategy {
  return Object.freeze({ kind: "swarm", reducer });
}

export function teamStrategyName(strategy: TeamStrategy): string {
  return strategy.kind;
}

export function teamStrategyDescription(strategy: TeamStrategy): string {
  switch (strategy.kind) {
    case "manager_subagent": return "Manager plans typed tasks, Subagents execute them, and the manager synthesizes";
    case "pipeline": return "Subagents execute in sequence";
    case "debate": return "Debaters propose independently and a judge synthesizes";
    case "swarm": return "Subagents inspect independently and a reducer synthesizes";
  }
}

File: sdks/python/src/echo_agent_sdk/team_strategy_values.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


class TeamStrategyKind(str, Enum):
    MANAGER_SUBAGENT = "manager_subagent"
    PIPELINE = "pipeline"
    DEBATE = "debate"
    SWARM = "swarm"


@dataclass(frozen=True, slots=True)
class TeamStrategy:
    kind: TeamStrategyKind
    members: tuple[str, ...] = ()
    judge: str | None = None
    debaters: tuple[str, ...] = ()
    reducer: str | None = None

    @classmethod
    def manager(cls) -> TeamStrategy:
        return cls(TeamStrategyKind.MANAGER_SUBAGENT)

    @classmethod
    def pipeline(cls, members: list[str] | tuple[str, ...]) -> TeamStrategy:
        return cls(TeamStrategyKind.PIPELINE, members=tuple(members))

    @classmethod
    def debate(cls, judge: str, debaters: list[str] | tuple[str, ...]) -> TeamStrategy:
        return cls(TeamStrategyKind.DEBATE, judge=judge, debaters=tuple(debaters))

    @classmethod
    def swarm(cls, reducer: str) -> TeamStrategy:
        return cls(TeamStrategyKind.SWARM, reducer=reducer)

    def name(self) -> str:
        return self.kind.value

    def description(self) -> str:
        return {
            TeamStrategyKind.MANAGER_SUBAGENT: "Manager plans typed tasks, Subagents execute them, and the manager synthesizes",
            TeamStrategyKind.PIPELINE: "Subagents execute in sequence",
            TeamStrategyKind.DEBATE: "Debaters propose independently and a judge synthesizes",
            TeamStrategyKind.SWARM: "Subagents inspect independently and a reducer synthesizes",
        }[self.kind]
